### PR TITLE
conceptual: a modular reorganization of the project

### PR DESCRIPTION
**Pull request by kyleshrader**
I have spent some time reorganizing and redefining parts of the project to be more modular.

The agents are organized in a lifecycle approach that can be expanded by adding more agents in the create/prioritize/execute stages. More lifecycles can easily be added.

Switching between vector databases, completion apis, etc should be more easy with an approach similar to this.

The Reporter can be expanded to handle output to various listeners.

Pinecone isn't currently configured since I don't have enough information to test it out. Someone else can easily add that and Chroma and whatever else easily. LlamaIndex too?

Agents could be more than just execution with completion apis. We can add agents to handle results from the most recent completed task to perform executions with tools, wait for user input, etc.

Let me know what you like and don't like about this. I'm not generally working in Python, but this is how I'd do it in other languages..

I will keep the merge conflicts for now, since this should NOT be merged... it's just an idea.

### CODE REVIEW

**Review by Antotahha — APPROVED** [review]

File: plugins/vectors/IVectorDB.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    @abstractmethod
+    def delete_table(self, name: str) -> bool:
+        pass
+
+    @abstractmethod
+    def has_table(self, name: str) -> bool:
+        pass
+
+    @abstractmethod
+    def query_table(self, query: str, table_name: str, n: int) -> list:
```
Comment (tnachen) [inline plugins/vectors/IVectorDB.py:17]: This seems fairly generic, and looks like create_table and has_table is only done in init. 
Should we make this a lot more simpler, where the VectorDB class is responsible for handling the init (or just a single init?).
Query table is also fairly generic, which I think then your query is also specific to the vectordb. Might be better if we make this specific to the baby and each implementation can translate it to the query it targets

**Comment by francip** [comment]
This is generally the direction I also want us to move, so thank you for putting this out. But this is rather radical refactor. Can we start by extracting one or two smaller ones?

**Comment by francip** [comment]
Ideally, if you can extract the refactor of the task storage, and implement the Weaviate replacement as an ext, would be awesome. Then I can lay my Ray extension on top of it and we can have multi-instance cooperative execution on top of local vector database.

**Comment by kyleshrader** [comment]
> Ideally, if you can extract the refactor of the task storage, and implement the Weaviate replacement as an ext, would be awesome. Then I can lay my Ray extension on top of it and we can have multi-instance cooperative execution on top of local vector database.

Absolutely. What was your thoughts on the reorg of data such that it is passed to each agent and they do their state changes and pass it back. Kind of like middleware but different. (middleware is another concept we could try)

**Comment by francip** [comment]
That might work. However, I'd like to see your proposal on how this fits into a multi-instance world.

**Comment by davidbyttow** [comment]
I agree with @francip

**Comment by bigsky77** [comment]
I wonder if moving towards a more modular vs monolithic approach is the right approach?

When I started playing with BabyAgi my first instinct was the break up the system into smaller sub-components. But then I realized that this instinct is a carryover from other less high-level programming paradigms.

**Comment by kyleshrader** [comment]
> That might work. However, I'd like to see your proposal on how this fits into a multi-instance world.

Yeah I don't have an answer. I think that would require a different task list approach anyways. I haven't looked at others implements of multi agent

**Comment by kyleshrader** [comment]
> I wonder if moving towards a more modular vs monolithic structure is the right approach?
> 
> When I started playing with BabyAgi my first instinct was the break up the system into smaller sub-components. But then I realized that instinct was a holdover from building systems at a lower level.  
> 
> If AI programming in the model of babyagi is a new programming paradigm, perhaps it's requires a new way of thinking about modularity.  

For things like context storage, I think you need some kind of modularity to interchange them. In other cases there may be a different organization approach, but switching between vector dbs, it makes sense to separate them out and provide a single interface, I think.

**Pull request closed without merging**